activerecord/test/cases/migration/change_schema_test.rb

```diff
@@ -286,7 +286,7 @@ def test_add_column_with_timestamp_type
         elsif current_adapter?(:OracleAdapter)
           assert_equal "TIMESTAMP(6)", column.sql_type
         else
-          assert_equal connection.type_to_sql("datetime"), column.sql_type
+          assert_equal connection.type_to_sql("datetime(6)"), column.sql_type
```

Comment:
I'm not sure about this change, but it gets CI passing again

Reply:
This is concerning. Do you know you needed this change?

Reply:
No. I made the change to see if it would fail any other tests, but it didn't. I'm not familiar with these tests, and haven't had a chance to deep dive and figure out why it was required.

Reply:
I think the SQL generated on SQLite3 for the `t.column :foo, :timestamp` now includes the precision of 6, when before it didn't. We need to investigate why this happened before merging it. I think this is the right behavior, but it is a behavior change, so we need to make sure the previous behavior was indeed a bug.

Reply:
Got it. This is indeed the right behavior.

Now you moved the code to add new default precision to after `aliased_types` is called. So `:timestamp` becomes `:datetime` and the new default is used.

Reply:
For posterity, [here](https://buildkite.com/rails/rails/builds/89779#01836711-48c5-46ba-8f24-b354c91eed4d) is a link to the failing build.

I've figured out why it was failing, and it's a similar reason to why this bug existed:

The incorrect virtual datetime precision was because the default assignment happened before we finalized the type from `virtual` to `datetime`. Similarly, [alias resolution](https://github.com/sambostock/rails/blob/b85ee91633bbc157ceb571cb4d3e5861ca7c15b6/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb#L543) happened after default precision assignment, so `timestamp` type columns implemented as `datetime` were using `nil` precision as their default instead of `6`.

So
- if that behavior **is not expected**, then this PR introduces a bug, but
- if that behavior **is expected**, then this PR fixes both bugs.

Based on the description of #42297, I believe this is intended:

> Set precision 6 by default for datetime columns, this is the standard behaviour for timestamps https://github.com/rails/rails/pull/34970 it would good to extend this to all datetimes by default.

In fact, I think we have an opportunity to refactor and remove duplicate code in the `t.timestamps` path, since we apply the default precision to all `datetime` columns.